vm/src/stdlib/os.rs

```diff
@@ -1127,6 +1130,123 @@ pub(super) mod _os {
         OutputMode::String.process_path(curdir_inner(vm)?, vm)
     }
 
+    #[cfg(unix)]
+    #[derive(FromArgs)]
+    struct RegisterAtForkArgs {
+        #[pyarg(named, optional)]
+        before: OptionalArg<PyObjectRef>,
+        #[pyarg(named, optional)]
+        after_in_parent: OptionalArg<PyObjectRef>,
+        #[pyarg(named, optional)]
+        after_in_child: OptionalArg<PyObjectRef>,
+    }
+
+    #[cfg(unix)]
+    #[pyfunction]
+    fn register_at_fork(
+        args: RegisterAtForkArgs,
+        _ignored: KwArgs,
+        vm: &VirtualMachine,
+    ) -> PyResult<()> {
+        let mut found_arg: bool = false; // hack to find atleas one arg
+        match args.before {
+            OptionalArg::Present(before) => {
+                if !before.is_callable() {
+                    return Err(vm.new_type_error("Args must be callable".to_owned()));
+                }
+                vm.state.before_forkers.lock().push(before);
+                found_arg = true
+            }
+            OptionalArg::Missing => (),
+        };
+        match args.after_in_parent {
+            OptionalArg::Present(after_in_parent) => {
+                if !after_in_parent.is_callable() {
+                    return Err(vm.new_type_error("Args must be callable".to_owned()));
+                }
+                vm.state.after_forkers_parent.lock().push(after_in_parent);
+                found_arg = true
+            }
+
+            OptionalArg::Missing => (),
+        };
+        match args.after_in_child {
+            OptionalArg::Present(after_in_child) => {
+                if !after_in_child.is_callable() {
+                    return Err(vm.new_type_error("Args must be callable".to_owned()));
+                }
+                vm.state.after_forkers_child.lock().push(after_in_child);
+                found_arg = true
+            }
+            OptionalArg::Missing => (),
+        };
+        if !found_arg {
+            return Err(vm.new_type_error("At least one argument is required.".to_owned()));
+        }
+
+        Ok(())
+    }
+
+    #[cfg(unix)]
+    fn run_at_forkers(mut funcs: Vec<PyObjectRef>, reversed: bool, vm: &VirtualMachine) {
+        if !funcs.is_empty() {
+            if reversed {
+                funcs.reverse();
+            }
+            for func in funcs.into_iter() {
+                if let Err(e) = func.call((), vm) {
+                    let exit = e.fast_isinstance(vm.ctx.exceptions.system_exit);
+                    vm.run_unraisable(e, Some("Exception ignored in".to_owned()), func);
+                    if exit {
+                        // Do nothing!
+                    }
+                }
+            }
+        }
+    }
+
+    #[cfg(unix)]
+    fn py_os_before_fork(vm: &VirtualMachine) -> PyResult<()> {
+        let before_forkers: Vec<PyObjectRef> = std::mem::take(&mut *vm.state.before_forkers.lock());
```

Comment:
Is this not clone but take? So that second run of fork doesn't be affect by it?

Reply:
Actually you are right.
```python
import os
def a():
   print('a')
def b():
   print('b')
def c():
   os.register_at_fork(before=a)
   os.fork()
   os.register_at_fork(before=b)
   os.fork()
c()
```
yields
```
a
b
a
b
a
``` 
in cpython. I have remove take operation and now rustpython as same results.
This should be a test case in cpython. 
```

Reply:
cpython PR: https://github.com/python/cpython/pull/103759